### PR TITLE
Add Unicore parser

### PR DESCRIPTION
**Pull request by julianoes**
This adds a separate class to parse Unicore messages which is then included as part of the NMEA parser.

This brings support for the Unicore UM982 module and heading specifically for the dual antenna setup.

The PR adds a unit test that can currently be run by a separate cmake command but can probably ultimately be included in other unit tests.

For now, the commands to run it, are:

```
cmake -Bbuild -H.
cmake --build build && build/gps-parser-test
```

FYI @vincentpoont2

### CODE REVIEW

File: src/crc.h [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#pragma once
+
+#include <cstdint>
+
+uint32_t calculateBlockCRC32(uint32_t length, uint8_t *buffer, uint32_t crc);
```
Comment (bkueng) [inline src/crc.h:38]: Is there a more specific name you could use?

File: src/unicore.cpp [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+		}
+
+		break;
+
+	case State::GotHashtag:
+		if (c == '*') {
+			_state = State::GotStar;
+
+			// Make sure buffer is zero terminated.
+			_buffer[_buffer_pos] = '\0';
```
Comment (bkueng) [inline src/unicore.cpp:55]: `_buffer_pos` can be equal to `sizeof(_buffer)` here.
You can change the check below to `if (_buffer_pos >= sizeof(_buffer)-1) {`

File: src/unicore.cpp [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+bool UnicoreParser::isHeading() const
+{
+	const char header[] = "UNIHEADINGA";
+
+	if (strncmp(header, _buffer, strlen(header)) != 0) {
+		return false;
+	}
+
+	return true;
```
Comment (bkueng) [inline src/unicore.cpp:124]: `return strncmp(header, _buffer, strlen(header)) == 0;`

File: src/unicore.cpp [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+	if (ptr == NULL) {
+		return false;
+	}
+
+	ptr = strtok(ptr, ",");
+
+	unsigned i = 0;
+
+	while (ptr != NULL) {
+		ptr = strtok(NULL, ",");
```
Comment (bkueng) [inline src/unicore.cpp:143]: This needs another check for `ptr != NULL`

File: src/unicore.cpp [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+		} else if (i == 2) {
+			_heading.heading_deg = strtof(ptr, NULL);
+			break;
+		}
+
+		++i;
+	}
+
+	return true;
```
Comment (bkueng) [inline src/unicore.cpp:158]: `return i >= 5;`

**Review by bkueng — COMMENTED** [review]
Looks good. Is only the heading a custom message, or does it also support a completely custom protocol? I'm just wondering if a completely separate driver would make sense.

> Stack is raised because we allocate 200 bytes for the Unicore parser.

The object is actually on the heap.

**Comment by julianoes** [comment]
Thanks for the review @bkueng, as always spot on!

> Is only the heading a custom message, or does it also support a completely custom protocol? I'm just wondering if a completely separate driver would make sense.

The Unicore sentences are sprinkled in-between the NMEA messages, so it's still primarily NMEA. If it was a separate driver, we would have to duplicate the NMEA part, or extract NMEA parsing functions and re-use them. I would vote to do it the way it is now.

> The object is actually on the heap.

You are right, I'll revert that.

File: src/crc.h (hunk 1 above)
Comment (julianoes) [reply src/crc.h:38]: I just extracted it out of the femtomes driver. But you're right, I might as well document and name it properly.

File: src/unicore.cpp (hunk 2 above)
Comment (julianoes) [reply src/unicore.cpp:55]: Oh, yes!

File: src/unicore.cpp (hunk 3 above)
Comment (julianoes) [reply src/unicore.cpp:124]: That would be too nice :)

File: src/unicore.cpp (hunk 4 above)
Comment (julianoes) [reply src/unicore.cpp:143]: Good catch!

File: src/unicore.cpp (hunk 5 above)
Comment (julianoes) [reply src/unicore.cpp:158]: Actually, I was going to just early return when i == 5 was parsed.

**Review by bkueng — APPROVED** [review]
> The Unicore sentences are sprinkled in-between the NMEA messages, so it's still primarily NMEA. If it was a separate driver, we would have to duplicate the NMEA part, or extract NMEA parsing functions and re-use them. I would vote to do it the way it is now.

Alright, fine with me. It's well separated.

**Pull request merged**